mlir/lib/Gradient/Transforms/GradMethods/HybridGradient.cpp

```diff
@@ -158,35 +162,73 @@ func::FuncOp genFullGradFunction(PatternRewriter &rewriter, Location loc, GradOp
 
                         Type gradientType = gradient.getType();
-                        auto rankGradient = gradientType.cast<RankedTensorType>().getRank();
-
-                        // sizes
-                        std::vector<int64_t> sizesSlice{shapeResult};
-                        for (int64_t sliceIndex = rankResult - 1; sliceIndex >= rankGradient;
-                             sliceIndex--) {
-                            sizesSlice[sliceIndex] = 1;
+                        if (auto gradientTensorType = dyn_cast<RankedTensorType>(gradientType)) {
+                            int64_t rankGradient = gradientTensorType.getRank();
+                            // sizes
+                            std::vector<int64_t> sizesSlice{shapeResult};
+                            for (int64_t sliceIndex = rankResult - 1; sliceIndex >= rankGradient;
+                                 sliceIndex--) {
+                                sizesSlice[sliceIndex] = 1;
+                            }
+
+                            // offset
+                            auto offsetSlice = allOffsets[index];
+                            for (int64_t offsetIndex = 0; offsetIndex < rankGradient;
+                                 offsetIndex++) {
+                                offsetSlice.insert(offsetSlice.begin(), 0);
+                            }
+                            result = rewriter.create<tensor::InsertSliceOp>(
+                                loc, resultType, gradient, result, ValueRange{}, ValueRange{},
+                                ValueRange{}, offsetSlice, sizesSlice, stridesSlice);
                         }
-
-                        // offset
-                        auto offsetSlice = allOffsets[index];
-                        for (int64_t offsetIndex = 0; offsetIndex < rankGradient; offsetIndex++) {
-                            offsetSlice.insert(offsetSlice.begin(), 0);
+                        else {
+                            assert(isa<FloatType>(gradient.getType()));
+                            SmallVector<Value> insertIndices;
+                            for (int64_t offset : allOffsets[index]) {
+                                insertIndices.push_back(
+                                    rewriter.create<index::ConstantOp>(loc, offset));
+                            }
+                            result = rewriter.create<tensor::InsertOp>(loc, gradient, result,
+                                                                       insertIndices);
                         }
-                        result = rewriter.create<tensor::InsertSliceOp>(
-                            loc, resultType, gradient, result, ValueRange{}, ValueRange{},
-                            ValueRange{}, offsetSlice, sizesSlice, stridesSlice);
                     }
                     hybridGradients.push_back(result);
                 }
             }
             else {
+                // Co-tangent is a rank 1 tensor
```

Comment:
Should the cotagent still be rank 1 when the result is rank 0?

Reply:
Yes, specifically in the context of our current hybrid gradient architecture, the quantum co-tangent always has shape `[<param_count>, ...<num_results>]`. This translates to a `tensor<?xf64>` when the result is rank 0, which is the case of a single measurement (regardless of if the measurement is a `tensor<f64>` or `f64`).

Reply:
>  the quantum co-tangent always has shape [<param_count>, ...<num_results>]

I thought the co-tangents are a tuple of size `<num_results>` where each element has shape `(<param_count>,)`?

But you are right this means the co-tangent is always rank 1 then! But that's because the result of the function we differentiate is always of rank 1 (the argmap function is what we differentiate).

I think it's this comment further down that confuses me:
> If the rank of the cotangent is 1, this implies the primal function returns a rank-0 value (either a scalar or a tensor<scalar>).

Reply:
> I thought the co-tangents are a tuple of size `<num_results>` where each element has shape `(<param_count>,)`?

Right, sorry, this is correct. I meant that each member of the co-tangents tuple has shape `[<param_count>, ...<cotangents[i].shape>]`. The name `num_results` was poorly chosen 😅 


> I think it's this comment further down that confuses me

So a primal function that returns `(f64, tensor<2xf64>, tensor<f64>)` will have cotangents `(tensor<?xf64>, tensor<?x2xf64>, tensor<?xf64>)`. If the cotangent at index `i` will always be of rank 1 greater than its corresponding primal result. I can edit the comment to try to make that more clear.

Reply:
> So a primal function that returns (f64, tensor<2xf64>, tensor<f64>) will have cotangents (tensor<?xf64>, tensor<?x2xf64>, tensor<?xf64>). If the cotangent at index i will always be of rank 1 greater than its corresponding primal result.

This sounds strange to me. By primal do you mean the whole QNode or the argmap function?
Because no matter what return types the QNode has, the argmap function always returns `tensor<?xf64>`. Consequently, the cotangent vector is also always of type `tensor<?xf64>`. Why would the rank be one higher than the corresponding result? Maybe I'm missing something 😅 

If you are talking about a generic function (independent of the QNode/argmap scenario), then a function that returns `(f64, tensor<2xf64>, tensor<f64>)` should have cotangents like `(_, tensor<2xf64>, tensor<f64>)` (I think the first one is omitted because Enzyme doesn't accept a cotangent for a scalar result right?).

Reply:
I'm specifically talking about the QNode/argmap situation, where "primal result" means one or more measurements. In order to compute the overall hybrid Jacobian, we need to run a backpropagation computation for each row of the Jacobian and there are as many rows as scalar element outputs of the hybrid circuit.

For example, if we have a hybrid function that returns `qml.expval(...), qml.probs()` with 1 wire, the QNode return would be `(tensor<f64>, tensor<2xf64>)`. There is thus one row of the hybrid Jacobian matrix (meaning a cotangent/quantum gradient vector of `tensor<?xf64>`) from the `expval`, and two rows of the hybrid Jacobian: one from each element of the `probs`.

As each row of the Jacobian matrix has a different quantum gradient vector of size `tensor<?xf64>`, the `expval` results in a quantum gradient of `tensor<?xf64>` while the `probs` results in a (transposed in the current implementation) quantum Jacobian of `tensor<?x2xf64>`. For both, the rank of the quantum gradient/Jacobian is 1 greater than the corresponding measurement.

I realize the terminology is confusing (which is a common problem in AD literature 😢). It might help to exclusively specify "quantum gradient" instead of "cotangent" because it's a cotangent with respect to the argmap, but not the circuit. This also only applies to the current hybrid gradient architecture.

(For completeness, the cotangents with respect to the circuit in this case would be `(1, [0, 0])`, `(0, [1, 0])`, `(0, [0, 1])`. It looks like the qgrad function uses these implicitly).

Reply:
> I realize the terminology is confusing (which is a common problem in AD literature 😢). It might help to exclusively specify "quantum gradient" instead of "cotangent" because it's a cotangent with respect to the argmap, but not the circuit. This also only applies to the current hybrid gradient architecture.

Ah got it, yes I think calling the quantum Jacobian the cotangent is the confusing part here, imo they are two separate entities. We do use individual rows of the quantum Jacobian (rank measurement + 1) as contangents, but the contangent itself only exists with respect to the argmap function (always fixed rank 1)😅